File: ios/Classes/coins/BTC.swift
/*
 BTC
 */

import WalletCore

class BTC: Coin  {
    init() {
        super.init(name: "BTC", coinType: .bitcoin)
    }
    
    override func generateAddress(path: String, mnemonic: String, passphrase: String) -> [String : String]? {
        let privateKey =  HDWallet(mnemonic: mnemonic, passphrase: passphrase)?.getKey(coin: self.coinType, derivationPath: path)
        if privateKey != nil {
            let publicKey = privateKey!.getPublicKeySecp256k1(compressed: true)
            let legacyAddress = BitcoinAddress(publicKey: publicKey, prefix: 0)
            let scriptHashAddress = BitcoinAddress(publicKey: publicKey, prefix: 5)
            return ["legacy": legacyAddress!.description,
                    "segwit": self.coinType.deriveAddress(privateKey: privateKey!),
                    "p2sh": scriptHashAddress!.description,
            ]
        }else {
            return nil
        }
        
    }
    
    override func signTransaction(path: String, txData: [String : Any], mnemonic: String, passphrase: String) -> String? {
        let privateKey = HDWallet(mnemonic: mnemonic, passphrase: passphrase)?.getKey(coin: self.coinType, derivationPath: path)
        let utxos: [[String: Any]] = txData["utxos"] as! [[String: Any]]
        var unspent: [BitcoinUnspentTransaction] = []
        
        if privateKey != nil {
            for utx in utxos {
                unspent.append(BitcoinUnspentTransaction.with {
                    $0.outPoint.hash = Data.reverse(hexString: utx["txid"] as! String)
                    $0.outPoint.index = utx["vout"] as! UInt32
                    $0.outPoint.sequence = UINT32_MAX
                    $0.amount = utx["value"] as! Int64
                    $0.script = Data(hexString: utx["script"] as! String)!
                })
            }
            let input: BitcoinSigningInput = BitcoinSigningInput.with {
                $0.hashType = BitcoinScript.hashTypeForCoin(coinType: .bitcoin)
                $0.amount = txData["amount"] as! Int64
                $0.toAddress = txData["toAddress"] as! String
                $0.changeAddress = txData["changeAddress"] as! String // can be same sender address
                $0.privateKey = [privateKey!.data]
                $0.plan = BitcoinTransactionPlan.with {
                    $0.amount = txData["amount"] as! Int64
                    $0.fee = txData["fees"] as! Int64 
                    $0.change = txData["change"] as! Int64
                    $0.utxos = unspent
                }
            }
            
            let output: BitcoinSigningOutput = AnySigner.sign(input: input, coin: .bitcoin)
            return output.encoded.hexString
        }else {
            return nil
        }
        
    }


    
    override func multiSignTransaction(txData: [String : Any], privateKeys: [String]) -> String? {
        let utxos: [[String: Any]] = txData["utxos"] as! [[String: Any]]
        var unspent: [BitcoinUnspentTransaction] = []
        
            for utx in utxos {
                unspent.append(BitcoinUnspentTransaction.with {
                    $0.outPoint.hash = Data.reverse(hexString: utx["txid"] as! String)
                    $0.outPoint.index = utx["vout"] as! UInt32
                    $0.outPoint.sequence = UINT32_MAX
                    $0.amount = utx["value"] as! Int64
                    $0.script = Data(hexString: utx["script"] as! String)!
                })
            }
            let privateKeyDataArray = privateKeys.compactMap { privateKey in
                return Data(hexString: privateKey)
            }

            let input: BitcoinSigningInput = BitcoinSigningInput.with {
                $0.hashType = BitcoinScript.hashTypeForCoin(coinType: .bitcoin)
                $0.amount = txData["amount"] as! Int64
                $0.toAddress = txData["toAddress"] as! String
                $0.changeAddress = txData["changeAddress"] as! String // can be same sender address
                $0.privateKey = privateKeyDataArray
                $0.plan = BitcoinTransactionPlan.with {
                    $0.amount = txData["amount"] as! Int64
                    $0.fee = txData["fees"] as! Int64
                    $0.change = txData["change"] as! Int64
                    $0.utxos = unspent
                }
            }
            
            let output: BitcoinSigningOutput = AnySigner.sign(input: input, coin: .bitcoin)
            return output.encoded.hexString
        
    }

}

File: android/src/main/kotlin/africa/ejara/trustdart/coins/BTC.kt
import africa.ejara.trustdart.Coin
import africa.ejara.trustdart.Numeric
import africa.ejara.trustdart.utils.toHex
import africa.ejara.trustdart.utils.toLong
import com.google.protobuf.ByteString
import wallet.core.java.AnySigner
import wallet.core.jni.BitcoinAddress
import wallet.core.jni.BitcoinScript
import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet
import wallet.core.jni.proto.Bitcoin

 


class BTC : Coin("BTC", CoinType.BITCOIN) {

    override fun generateAddress(
        path: String,
        mnemonic: String,
        passphrase: String
    ): Map<String, String>? {
        val wallet = HDWallet(mnemonic, passphrase)
        val privateKey = wallet.getKey(coinType, path)
        val publicKey = privateKey.getPublicKeySecp256k1(true)
        val address = BitcoinAddress(publicKey, coinType!!.p2pkhPrefix())
        return mapOf(
            "legacy" to address.description(),
            "segwit" to coinType!!.deriveAddress(privateKey)
        )
    }

    override fun signTransaction(
        path: String,
        txData: Map<String, Any>,
        mnemonic: String,
        passphrase: String
    ): String? {
        val wallet = HDWallet(mnemonic, passphrase)
        val privateKey = wallet.getKey(coinType, path)
        val utxos: List<Map<String, Any>> = txData["utxos"] as List<Map<String, Any>>

        val input = Bitcoin.SigningInput.newBuilder()
            .setAmount(txData["amount"]!!.toLong())
            .setHashType(BitcoinScript.hashTypeForCoin(coinType))
            .setToAddress(txData["toAddress"] as String)
            .setChangeAddress(txData["changeAddress"] as String)
            .setByteFee(1)
            .addPrivateKey(ByteString.copyFrom(privateKey.data()))

        for (utx in utxos) {
            val txHash = Numeric.hexStringToByteArray(utx["txid"] as String)
            txHash.reverse()
            val outPoint = Bitcoin.OutPoint.newBuilder()
                .setHash(ByteString.copyFrom(txHash))
                .setIndex(utx["vout"] as Int)
                .setSequence(Long.MAX_VALUE.toInt())
                .build()
            val txScript = Numeric.hexStringToByteArray(utx["script"] as String)
            val utxo = Bitcoin.UnspentTransaction.newBuilder()
                .setAmount(utx["value"]!!.toLong())
                .setOutPoint(outPoint)
                .setScript(ByteString.copyFrom(txScript))
                .build()
            input.addUtxo(utxo)
        }

        var output = AnySigner.sign(input.build(), coinType, Bitcoin.SigningOutput.parser())
        // since we want to set our own fee
        // but such functionality is not obvious in the trustwalletcore library
        // a hack is used for now to calculate the byteFee
        val size = output.encoded.toByteArray().size
        val fees = txData["fees"]!!.toLong()
        if (size > 0) { // prevent division by zero
            val byteFee = fees.div(size) // this gives the fee per byte truncated to Long
            // now we set new byte size
            if (byteFee > 1) input.byteFee = byteFee
        }
        output = AnySigner.sign(input.build(), coinType, Bitcoin.SigningOutput.parser())
        return Numeric.toHexString(output.encoded.toByteArray())
    }


    override fun multiSignTransaction(
        txData: Map<String, Any>,
        privateKeys: ArrayList<String>
    ): String? {
        val utxos: List<Map<String, Any>> = txData["utxos"] as List<Map<String, Any>>

        val input = Bitcoin.SigningInput.newBuilder()
            .setAmount(txData["amount"]!!.toLong())
            .setHashType(BitcoinScript.hashTypeForCoin(coinType))
            .setToAddress(txData["toAddress"] as String)
            .setChangeAddress(txData["changeAddress"] as String)
            .setByteFee(1)

        val byteStrings: MutableList<ByteString> = privateKeys.map { ByteString.copyFrom(Numeric.hexStringToByteArray(it)) }.toMutableList()

        input.addAllPrivateKey(byteStrings);

        for (utx in utxos) {
            val txHash = Numeric.hexStringToByteArray(utx["txid"] as String)
            txHash.reverse()
            val outPoint = Bitcoin.OutPoint.newBuilder()
                .setHash(ByteString.copyFrom(txHash))
                .setIndex(utx["vout"] as Int)
                .setSequence(Long.MAX_VALUE.toInt())
                .build()
            val txScript = Numeric.hexStringToByteArray(utx["script"] as String)
            val utxo = Bitcoin.UnspentTransaction.newBuilder()
                .setAmount(utx["value"]!!.toLong())
                .setOutPoint(outPoint)
                .setScript(ByteString.copyFrom(txScript))
                .build()
            input.addUtxo(utxo)
        }

        var output = AnySigner.sign(input.build(), coinType, Bitcoin.SigningOutput.parser())

        // since we want to set our own fee
        // but such functionality is not obvious in the trustwalletcore library
        // a hack is used for now to calculate the byteFee
        val size = output.encoded.toByteArray().size
        val fees = txData["fees"]!!.toLong()
        if (size > 0) { // prevent division by zero
            val byteFee = fees.div(size) // this gives the fee per byte truncated to Long
            // now we set new byte size
            if (byteFee > 1) input.byteFee = byteFee
        }
        output = AnySigner.sign(input.build(), coinType, Bitcoin.SigningOutput.parser())
        return Numeric.toHexString(output.encoded.toByteArray())
    }

};
           